File: android/src/main/java/expo/modules/tuyasdk/DeviceManagementModule.kt
package expo.modules.tuyasdk

import com.thingclips.smart.device.sharedevice.sdk.ThingShareBizKit
import com.thingclips.smart.device.sharedevice.sdk.bean.ShareMember
import com.thingclips.smart.home.sdk.bean.DeviceShareBean
import com.thingclips.smart.home.sdk.bean.ShareReceivedUserDetailBean
import com.thingclips.smart.home.sdk.bean.SharedUserInfoBean
import com.thingclips.smart.home.sdk.callback.IThingResultCallback
import com.thingclips.smart.sdk.api.IResultCallback
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class DeviceManagementModule : Module() {

        private val shareManager
                get() = ThingShareBizKit.getThingDeviceShareManager()

        override fun definition() = ModuleDefinition {
                Name("ExpoTuyaDeviceManagement")

                // ─── Sharer APIs ─────────────────────────────────────────────────────

                AsyncFunction("supportShare") { resId: String, resType: Int, promise: Promise ->
                        shareManager.isSupportDeviceShare(
                                resId,
                                resType,
                                object : IThingResultCallback<Boolean> {
                                        override fun onSuccess(result: Boolean) =
                                                promise.resolve(result)
                                        override fun onError(code: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_SUPPORT_SHARE",
                                                        msg ?: "supportShare failed",
                                                        null
                                                )
                                }
                        )
                }

                AsyncFunction("remainingShareTimes") { resId: String, resType: Int, promise: Promise
                        ->
                        // Not exposed in this SDK version — return -1 (unlimited)
                        promise.resolve(-1)
                }

                AsyncFunction("share") {
                        resId: String,
                        resType: Int,
                        spaceId: Double,
                        userAccount: String,
                        promise: Promise ->
                        shareManager.shareToUser(
                                resId,
                                resType,
                                spaceId.toLong(),
                                userAccount,
                                object : IThingResultCallback<SharedUserInfoBean> {
                                        override fun onSuccess(result: SharedUserInfoBean) {
                                                promise.resolve(
                                                        mapOf(
                                                                "memberId" to result.memeberId,
                                                                "nickName" to
                                                                        (result.remarkName ?: ""),
                                                                "userName" to
                                                                        (result.userName ?: "")
                                                        )
                                                )
                                        }
                                        override fun onError(code: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_SHARE",
                                                        msg ?: "share failed",
                                                        null
                                                )
                                }
                        )
                }

                AsyncFunction("removeReceiver") {
                        memberId: Double,
                        resId: String,
                        resType: Int,
                        promise: Promise ->
                        shareManager.removeReceiver(
                                memberId.toLong(),
                                resId,
                                resType,
                                object : IResultCallback {
                                        override fun onSuccess() = promise.resolve(null)
                                        override fun onError(code: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_REMOVE_RECEIVER",
                                                        msg ?: "removeReceiver failed",
                                                        null
                                                )
                                }
                        )
                }

                AsyncFunction("receivers") {
                        resId: String,
                        resType: Int,
                        page: Int,
                        pageSize: Int,
                        promise: Promise ->
                        shareManager.getReceivers(
                                resId,
                                resType,
                                page,
                                pageSize,
                                object : IThingResultCallback<List<ShareMember>> {
                                        override fun onSuccess(result: List<ShareMember>) =
                                                promise.resolve(result.map { shareMemberToMap(it) })
                                        override fun onError(code: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_RECEIVERS",
                                                        msg ?: "receivers failed",
                                                        null
                                                )
                                }
                        )
                }

                AsyncFunction("updateShareExpirationDate") {
                        memberId: Double,
                        resId: String,
                        resType: Int,
                        mode: Int,
                        endTime: Double,
                        promise: Promise ->
                        // Not exposed in this SDK version
                        promise.resolve(null)
                }

                AsyncFunction("relationMembers") { promise: Promise ->
                        shareManager.getRelationMembers(
                                object : IThingResultCallback<List<ShareMember>> {
                                        override fun onSuccess(result: List<ShareMember>) =
                                                promise.resolve(result.map { shareMemberToMap(it) })
                                        override fun onError(code: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_RELATION_MEMBERS",
                                                        msg ?: "relationMembers failed",
                                                        null
                                                )
                                }
                        )
                }

                AsyncFunction("removeRelationMember") { uid: String, promise: Promise ->
                        // Not exposed in this SDK version
                        promise.resolve(null)
                }

                AsyncFunction("createShareInfo") {
                        resId: String,
                        resType: Int,
                        spaceId: Double,
                        shareType: Int,
                        shareCount: Int,
                        promise: Promise ->
                        // Not exposed in this SDK version
                        promise.resolve(mapOf("content" to "", "code" to "", "shortUrl" to ""))
                }

                // ─── Receiver APIs ───────────────────────────────────────────────────

                AsyncFunction("validateShareCode") { code: String, promise: Promise ->
                        shareManager.parseShortLinkAvailability(
                                code,
                                object : IThingResultCallback<Boolean> {
                                        override fun onSuccess(result: Boolean) =
                                                promise.resolve(result)
                                        override fun onError(code2: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_VALIDATE_CODE",
                                                        msg ?: "validateShareCode failed",
                                                        null
                                                )
                                }
                        )
                }

                AsyncFunction("shareCodeInfo") { code: String, promise: Promise ->
                        // Not exposed in this SDK version
                        promise.resolve(
                                mapOf(
                                        "appId" to "",
                                        "resId" to "",
                                        "resType" to 1,
                                        "resIcon" to "",
                                        "resName" to "",
                                        "nickName" to "",
                                        "shareSource" to 0,
                                        "groupId" to 0L
                                )
                        )
                }

                AsyncFunction("acceptShare") { code: String, promise: Promise ->
                        shareManager.acceptShare(
                                code,
                                object : IResultCallback {
                                        override fun onSuccess() = promise.resolve(null)
                                        override fun onError(code2: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_ACCEPT_SHARE",
                                                        msg ?: "acceptShare failed",
                                                        null
                                                )
                                }
                        )
                }

                AsyncFunction("removeShare") { resId: String, resType: Int, promise: Promise ->
                        shareManager.removeReceivedShare(
                                resId,
                                resType,
                                object : IResultCallback {
                                        override fun onSuccess() = promise.resolve(null)
                                        override fun onError(code: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_REMOVE_SHARE",
                                                        msg ?: "removeShare failed",
                                                        null
                                                )
                                }
                        )
                }

                AsyncFunction("sharers") { promise: Promise ->
                        shareManager.getShareReceivedUserList(
                                object : IThingResultCallback<List<SharedUserInfoBean>> {
                                        override fun onSuccess(result: List<SharedUserInfoBean>) {
                                                promise.resolve(
                                                        result.map {
                                                                mapOf(
                                                                        "memberId" to it.memeberId,
                                                                        "nickName" to
                                                                                (it.remarkName
                                                                                        ?: ""),
                                                                        "userName" to
                                                                                (it.userName ?: "")
                                                                )
                                                        }
                                                )
                                        }
                                        override fun onError(code: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_SHARERS",
                                                        msg ?: "sharers failed",
                                                        null
                                                )
                                }
                        )
                }

                AsyncFunction("sharerName") { resId: String, resType: Int, promise: Promise ->
                        shareManager.getSharerName(
                                resId,
                                resType,
                                object : IThingResultCallback<String> {
                                        override fun onSuccess(result: String) =
                                                promise.resolve(result)
                                        override fun onError(code: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_SHARER_NAME",
                                                        msg ?: "sharerName failed",
                                                        null
                                                )
                                }
                        )
                }

                AsyncFunction("sharerDetail") { memberId: Double, promise: Promise ->
                        shareManager.getSharerInfoDetail(
                                memberId.toLong(),
                                object : IThingResultCallback<ShareReceivedUserDetailBean> {
                                        override fun onSuccess(
                                                result: ShareReceivedUserDetailBean
                                        ) {
                                                val devices =
                                                        result.devices?.map { d: DeviceShareBean ->
                                                                mapOf(
                                                                        "devId" to (d.devId ?: ""),
                                                                        "name" to
                                                                                (d.deviceName
                                                                                        ?: ""),
                                                                        "iconUrl" to
                                                                                (d.iconUrl ?: "")
                                                                )
                                                        }
                                                                ?: emptyList()
                                                promise.resolve(
                                                        mapOf(
                                                                "devices" to devices,
                                                                "mobile" to (result.mobile ?: ""),
                                                                "name" to
                                                                        (result.nameWithoutRemark
                                                                                ?: ""),
                                                                "remarkName" to
                                                                        (result.remarkName ?: "")
                                                        )
                                                )
                                        }
                                        override fun onError(code: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_SHARER_DETAIL",
                                                        msg ?: "sharerDetail failed",
                                                        null
                                                )
                                }
                        )
                }

                AsyncFunction("removeSharer") { memberId: Double, promise: Promise ->
                        shareManager.removeReceivedUserShare(
                                memberId.toLong(),
                                object : IResultCallback {
                                        override fun onSuccess() = promise.resolve(null)
                                        override fun onError(code: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_REMOVE_SHARER",
                                                        msg ?: "removeSharer failed",
                                                        null
                                                )
                                }
                        )
                }

                AsyncFunction("updateSharer") { memberId: Double, name: String, promise: Promise ->
                        shareManager.renameReceivedShareNickname(
                                memberId.toLong(),
                                name,
                                object : IResultCallback {
                                        override fun onSuccess() = promise.resolve(null)
                                        override fun onError(code: String?, msg: String?) =
                                                promise.reject(
                                                        "ERR_UPDATE_SHARER",
                                                        msg ?: "updateSharer failed",
                                                        null
                                                )
                                }
                        )
                }
        }

        // ─── Helpers ─────────────────────────────────────────────────────────────

        private fun shareMemberToMap(m: ShareMember): Map<String, Any?> =
                mapOf(
                        "memberId" to (m.memberId ?: 0L),
                        "nickName" to (m.nickName ?: ""),
                        "userName" to (m.userName ?: ""),
                        "iconUrl" to (m.iconUrl ?: ""),
                        "shareMode" to m.shareMode,
                        "endTime" to (m.endTime ?: 0L),
                        "uid" to (m.uid ?: "")
                )
}

File: ios/DeviceManagementModule.swift
import ExpoModulesCore
import ThingSmartBusinessExtensionKit

public class DeviceManagementModule: Module {

    public func definition() -> ModuleDefinition {
        Name("ExpoTuyaDeviceManagement")

        // ─── Sharer APIs ─────────────────────────────────────────────────────

        AsyncFunction("supportShare") { (resId: String, resType: Int32, promise: Promise) in
            ThingDeviceShareManager.supportShare(
                resId,
                resType: ThingShareResType(rawValue: Int(resType)) ?? .device,
                success: { promise.resolve($0) },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("remainingShareTimes") { (resId: String, resType: Int32, promise: Promise) in
            ThingDeviceShareManager.remainingShareTimes(
                resId,
                resType: ThingShareResType(rawValue: Int(resType)) ?? .device,
                success: { promise.resolve(Int($0)) },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("share") { (resId: String, resType: Int32, spaceId: Int64, userAccount: String, promise: Promise) in
            ThingDeviceShareManager.share(
                resId,
                resType: ThingShareResType(rawValue: Int(resType)) ?? .device,
                spaceId: spaceId,
                userAccount: userAccount,
                success: { result in
                    promise.resolve([
                        "memberId": result.memberId,
                        "nickName": result.nickName,
                        "userName": result.userName,
                    ])
                },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("removeReceiver") { (memberId: Int, resId: String, resType: Int32, promise: Promise) in
            ThingDeviceShareManager.removeReceiver(
                memberId,
                resId: resId,
                resType: ThingShareResType(rawValue: Int(resType)) ?? .device,
                success: { promise.resolve(nil) },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("receivers") { (resId: String, resType: Int32, page: Int32, pageSize: Int32, promise: Promise) in
            ThingDeviceShareManager.receivers(
                resId,
                resType: ThingShareResType(rawValue: Int(resType)) ?? .device,
                page: page,
                pageSize: pageSize,
                success: { members in
                    promise.resolve((members ?? []).map { self.shareMemberToDict($0) })
                },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("updateShareExpirationDate") { (memberId: Int, resId: String, resType: Int32, mode: Int32, endTime: Int64, promise: Promise) in
            ThingDeviceShareManager.updateShareExpirationDate(
                memberId,
                resId: resId,
                resType: ThingShareResType(rawValue: Int(resType)) ?? .device,
                mode: ThingShareValidationType(rawValue: Int(mode)) ?? .forever,
                endTime: endTime,
                success: { promise.resolve(nil) },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("relationMembers") { (promise: Promise) in
            ThingDeviceShareManager.relationMembers(
                { members in
                    promise.resolve((members ?? []).map { self.shareMemberToDict($0) })
                },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("removeRelationMember") { (uid: String, promise: Promise) in
            ThingDeviceShareManager.removeRelationMember(
                uid,
                success: { promise.resolve(nil) },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("createShareInfo") { (resId: String, resType: Int32, spaceId: Int64, shareType: Int32, shareCount: Int32, promise: Promise) in
            ThingDeviceShareManager.createShareInfo(
                resId,
                resType: ThingShareResType(rawValue: Int(resType)) ?? .device,
                spaceId: spaceId,
                shareType: ThingDeviceShareType(rawValue: Int(shareType)) ?? .account,
                shareCount: shareCount,
                success: { info in
                    promise.resolve([
                        "content": info.content,
                        "code": info.code,
                        "shortUrl": info.shortUrl,
                    ])
                },
                failure: { promise.reject($0) }
            )
        }

        // ─── Receiver APIs ───────────────────────────────────────────────────

        AsyncFunction("validateShareCode") { (code: String, promise: Promise) in
            ThingDeviceShareManager.validate(
                code,
                success: { promise.resolve($0) },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("shareCodeInfo") { (code: String, promise: Promise) in
            ThingDeviceShareManager.shareCodeInfo(
                code,
                success: { info in
                    promise.resolve([
                        "appId": info.appId,
                        "resId": info.resId,
                        "resType": info.resType.rawValue,
                        "resIcon": info.resIcon,
                        "resName": info.resName,
                        "nickName": info.nickName,
                        "shareSource": info.shareSource.rawValue,
                        "groupId": info.groupId,
                    ])
                },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("acceptShare") { (code: String, promise: Promise) in
            ThingDeviceShareManager.accept(
                code,
                success: { promise.resolve(nil) },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("removeShare") { (resId: String, resType: Int32, promise: Promise) in
            ThingDeviceShareManager.removeShare(
                resId,
                resType: ThingShareResType(rawValue: Int(resType)) ?? .device,
                success: { promise.resolve(nil) },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("sharers") { (promise: Promise) in
            ThingDeviceShareManager.sharers(
                { sharers in
                    let result = (sharers ?? []).map { s -> [String: Any] in
                        ["memberId": s.memberId, "nickName": s.nickName, "userName": s.userName]
                    }
                    promise.resolve(result)
                },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("sharerName") { (resId: String, resType: Int32, promise: Promise) in
            ThingDeviceShareManager.sharerName(
                resId,
                resType: ThingShareResType(rawValue: Int(resType)) ?? .device,
                success: { promise.resolve($0) },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("sharerDetail") { (memberId: Int, promise: Promise) in
            ThingDeviceShareManager.sharerDetail(
                memberId,
                success: { detail in
                    let devices = detail.devices.map { d -> [String: Any] in
                        ["devId": d.devId, "name": d.name, "iconUrl": d.iconUrl]
                    }
                    promise.resolve([
                        "devices": devices,
                        "mobile": detail.mobile,
                        "name": detail.name,
                        "remarkName": detail.remarkName,
                    ])
                },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("removeSharer") { (memberId: Int, promise: Promise) in
            ThingDeviceShareManager.removeSharer(
                memberId,
                success: { promise.resolve(nil) },
                failure: { promise.reject($0) }
            )
        }

        AsyncFunction("updateSharer") { (memberId: Int, name: String, promise: Promise) in
            ThingDeviceShareManager.updateSharer(
                memberId,
                name: name,
                success: { promise.resolve(nil) },
                failure: { promise.reject($0) }
            )
        }
    }

    // MARK: - Helpers

    private func shareMemberToDict(_ m: ThingSmartShareMemberModel) -> [String: Any] {
        return [
            "memberId": m.memberId,
            "nickName": m.nickName,
            "userName": m.userName,
            "iconUrl": m.iconUrl,
            "shareMode": m.shareMode.rawValue,
            "endTime": m.endTime,
            "uid": m.uid,
        ]
    }

    private func makeError(_ message: String) -> NSError {
        return NSError(
            domain: "ExpoTuyaDeviceManagement",
            code: -1,
            userInfo: [NSLocalizedDescriptionKey: message]
        )
    }
}
